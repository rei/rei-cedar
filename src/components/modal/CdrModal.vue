<script>
import { debounce } from 'lodash-es';
import tabbable from 'tabbable';
import {
  useCssModule, computed, ref, watch, onMounted, nextTick, onUnmounted, defineComponent,
} from 'vue';
import {
  CdrBreakpointSm, CdrSpaceOneX, CdrSpaceTwoX,
} from '@rei/cdr-tokens/dist/js/cdr-tokens.mjs';
import onTransitionEnd from './onTransitionEnd';
import CdrButton from '../button/CdrButton.vue';
import IconXLg from '../icon/comps/x-lg.vue';
import mapClasses from '../../utils/mapClasses';

/** Disruptive, action-blocking overlays used to display important information */
export default defineComponent({
  name: 'CdrModal',
  components: { CdrButton, IconXLg },
  props: {
    /**
     * Toggles the state of the modal
     * @demoIgnore true
     */
    opened: {
      type: Boolean,
      required: true,
    },
    /**
     * Sets `aria-label` and modal title text. Can also use title slot to set title.
     */
    label: {
      type: String,
      required: true,
    },
    /**
     * Toggles the modal title text, which comes from `label` prop or `title` slot.
     */
    showTitle: {
      type: Boolean,
      required: false,
      default: true,
    },
    /**
     * Text for aria-describedby attribute. Applied to modal content element
     */
    ariaDescribedby: {
      type: String,
      required: false,
      default: null,
    },
    /**
     * Sets the `role` attribute on the modal content element
     * @values dialog, alertDialog
     */
    role: {
      type: String,
      required: false,
      default: 'dialog',
    },
    /**
     * Sets unique `id` for modal
     */
    id: {
      type: String,
      required: false,
      default: null,
    },
    /** Adds custom class to the `cdr-modal__overlay` div */
    overlayClass: String,
    /** Adds custom class to the `cdr-modal__outerWrap` div */
    wrapperClass: String,
    /** Adds custom class to the `cdr-modal__innerWrap` div */
    contentClass: String,
    /** Sets duration for modal's close animation */
    animationDuration: {
      type: Number,
      default: 300,
    },
  },
  emits: {
    /** Fires when modal is closed */
    closed: null,
  },
  setup(props, ctx) {
    const baseClass = 'cdr-modal';
    let unsubscribe;
    let lastActive;
    const modalClosed = ref(!props.opened);
    const isOpening = ref(false);

    const offset = ref(null);
    const headerHeight = ref(0);
    const totalHeight = ref(0);
    const scrollHeight = ref(0);
    const offsetHeight = ref(0);
    const offsetWidth = ref(0);
    const clientWidth = ref(0);
    const fullscreen = ref(false);

    const modalEl = ref(null);
    const wrapperEl = ref(null);
    const contentEl = ref(null);
    const headerEl = ref(null);

    const measureContent = () => {
      nextTick(() => {
        totalHeight.value = window.innerHeight;
        fullscreen.value = window.innerWidth < CdrBreakpointSm;
        headerHeight.value = headerEl.value?.offsetHeight;
        scrollHeight.value = contentEl.value?.scrollHeight;
        offsetHeight.value = contentEl.value?.offsetHeight;
        offsetWidth.value = contentEl.value?.offsetWidth;
        clientWidth.value = contentEl.value?.clientWidth;
      });
    };

    const onClick = (e) => {
      ctx.emit('closed', e);
    };

    const handleKeyDown = ({ key }) => {
      switch (key) {
        case 'Escape':
        case 'Esc':
          onClick();
          break;
        default: break;
      }
    };
    const handleFocus = (e) => {
      const { documentElement } = document;
      if (modalEl.value.contains(e.target) || !documentElement) return;

      const tabbables = tabbable(documentElement);
      const these = tabbable(modalEl.value);
      const nextIx = tabbables.indexOf(e.target);
      const firstModalIx = tabbables.indexOf(these[0]);
      const nextRef = nextIx < firstModalIx ? these[these.length - 1] : these[0];
      if (nextRef) nextRef.focus();
    };

    const handleResize = debounce(() => {
      measureContent();
    }, 300);

    const addNoScroll = () => {
      const { documentElement, body } = document;
      offset.value = {
        x: window.scrollX
      || (documentElement || {}).scrollLeft
      || (body || {}).scrollLeft
      || 0,
        y: window.scrollY
      || (documentElement || {}).scrollTop
      || (body || {}).scrollTop
      || 0,
      };

      if (documentElement) {
        documentElement.classList.add('cdr-modal__noscroll');
        // keep current scroll position manually
        documentElement.style.top = `-${offset.value.y}px`;
        documentElement.style.left = `-${offset.value.x}px`;
      }

      if (body) {
        body.classList.add('cdr-modal__noscroll');
      }
    };

    const removeNoScroll = () => {
      const { documentElement, body } = document;

      if (body) {
        body.classList.remove('cdr-modal__noscroll');
      }

      if (documentElement) {
        documentElement.classList.remove('cdr-modal__noscroll');
        documentElement.style.top = '';
        documentElement.style.left = '';
      }
    };

    const addHandlers = () => {
      document.addEventListener('focusin', handleFocus, true);
      document.addEventListener('keydown', handleKeyDown);
    };

    const handleOpened = () => {
      const { activeElement } = document;
      addNoScroll();
      isOpening.value = true;
      modalClosed.value = false;
      lastActive = activeElement;

      nextTick(() => {
        if (modalEl.value) modalEl.value.focus(); // wrapped in if so testing error isn't thrown
        measureContent();
        addHandlers();

        setTimeout(() => {
          // for some reason Safari scrolls the wrapper down a bit?
          // doesn't work without setTimeout for some unknown reason
          if (wrapperEl.value) wrapperEl.value.scrollTop = 0;

          // there is a race condition for measuring overflow when modal defaults to open,
          // this seems to cover that
          measureContent();
        });
      });
    };

    const handleClosed = () => {
      const { documentElement } = document;
      document.removeEventListener('keydown', handleKeyDown);
      document.removeEventListener('focusin', handleFocus, true);
      isOpening.value = false;

      unsubscribe = onTransitionEnd(
        wrapperEl.value,
        () => {
          if (isOpening.value) return;
          unsubscribe();
          removeNoScroll();
          unsubscribe = null;
          modalClosed.value = true;

          // handle scroll-behavior: smooth
          if (documentElement) documentElement.style.scrollBehavior = 'auto';
          // restore previous scroll position
          window.scrollTo(offset.value.x, offset.value.y);
          if (documentElement) documentElement.style.scrollBehavior = '';

          if (lastActive) lastActive.focus();
        },
        props.animationDuration + 16,
      );
    };

    const dialogAttrs = computed(() => ({
      ...ctx.attrs,
      'aria-describedby': props.ariaDescribedby,
      'aria-modal': 'true',
      id: props.id,
    }));
    const verticalSpace = computed(() => {
      // contentWrap vertical padding
      const fullscreenSpace = Number(CdrSpaceTwoX);
      const windowedSpace = Number(CdrSpaceTwoX) + Number(CdrSpaceOneX);

      return fullscreen.value
        ? fullscreenSpace
        : windowedSpace + fullscreenSpace;
      // fullscreen, here, would account for outerWrap padding, which is the same CdrSpaceTwoX
    });
    const scrollMaxHeight = computed(() => totalHeight.value
    - headerHeight.value
    - verticalSpace.value);

    const scrollPadding = computed(() => {
      const isScrolling = scrollHeight.value > offsetHeight.value;
      const hasScrollbar = offsetWidth.value - clientWidth.value > 0;
      if (isScrolling && hasScrollbar) {
        return 4;
      } if (isScrolling) {
        return 12;
      }
      return 0;
    });

    watch(() => props.opened, (newValue, oldValue) => {
      if (!!newValue === !!oldValue) return;
      // eslint-disable-next-line no-unused-expressions
      newValue ? handleOpened() : handleClosed();
    });

    onMounted(() => {
      if (props.opened) {
        handleOpened();
      }
      window.addEventListener('resize', handleResize);
    });

    onUnmounted(() => {
      window.removeEventListener('resize', handleResize);
    });
    return {
      style: useCssModule(),
      mapClasses,
      baseClass,
      modalClosed,
      modalEl,
      wrapperEl,
      contentEl,
      headerEl,
      scrollMaxHeight,
      scrollPadding,
      onClick,
      dialogAttrs,

    };
  },

});

</script>

<template>
  <div
    :class="mapClasses(style, baseClass, !opened && 'cdr-modal--closed')"
    ref="wrapperEl"
  >
    <div
      :class="[style['cdr-modal__outerWrap'], wrapperClass]"
    >
      <div
        aria-hidden="true"
        @click="onClick"
        :class="[style['cdr-modal__overlay'], overlayClass]"
      />
      <!-- This div (and the one below) is used to avoid a screen reader "keyboard" trap where the
           content outside the modal is not properly obscured when opened using certain readers.
           Once more browsers catch up to the spec, this hack can probably be removed.
           https://a11ysupport.io/tests/apg__modal-dialog-example -->
      <div :tabIndex="opened ? '0' : undefined" />
      <div
        ref="modalEl"
        :class="mapClasses(style, 'cdr-modal__contentWrap', 'cdr-modal__dialog')"
        tabIndex="-1"
        :role="role"
        :aria-modal="!!opened"
        :aria-label="label"
        v-bind="dialogAttrs"
      >
        <!-- @slot Use to override the entire CdrModal content container.
          You must provide an explicit way to close the modal
          to meet UX and accessibility standards  -->
        <slot name="modal">
          <div
            :class="[style['cdr-modal__innerWrap'], contentClass]"
            :style="modalClosed ? {display: 'none'} : undefined"
          >
            <section>
              <div :class="style['cdr-modal__content']">
                <div
                  :class="style['cdr-modal__header']"
                  ref="headerEl"
                >
                  <div :class="style['cdr-modal__title']">
                    <!-- @slot Use to override the default title -->
                    <slot
                      name="title"
                      v-if="showTitle"
                    >
                      <h1>{{ label }}</h1>
                      <!-- @slot CdrModal content -->
                    </slot>
                  </div>
                  <cdr-button
                    :class="style['cdr-modal__close-button']"
                    icon-only
                    :with-background="true"
                    @click="onClick"
                    aria-label="Close"
                  >
                    <template #icon>
                      <icon-x-lg inherit-color />
                    </template>
                  </cdr-button>
                </div>

                <div
                  :class="style['cdr-modal__text-content']"
                  :style="{ maxHeight: `${scrollMaxHeight}px`, paddingRight: `${scrollPadding}px`}"
                  role="document"
                  ref="contentEl"
                  tabindex="0"
                >
                  <slot />
                </div>
              </div>
            </section>
          </div>

        </slot>
      </div>
      <div :tabIndex="opened ? '0' : undefined" />
    </div>
  </div>

</template>

<style lang="scss" module src="./styles/CdrModal.module.scss">
</style>
