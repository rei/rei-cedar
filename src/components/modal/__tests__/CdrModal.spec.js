import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrModal from '../CdrModal.vue';

describe('CdrModal.vue', () => {
  describe('default open', ()=>{
    let wrapper;
    let elem;
    beforeEach(()=>{
      elem = document.createElement('div')
      if (document.body) {
        document.body.appendChild(elem)
      }
      wrapper = mount(CdrModal, {
        propsData: {
          opened: true,
          label: "Label is the modal title"
        },
        slots: {
          default: 'Sticky content',
        },
        attachTo: elem,
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('contains the "cdr-modal__noscroll" class', () => {
      expect(document.documentElement.classList.contains('cdr-modal__noscroll')).toBe(true);
    });

    it('handleKeyDown', async () => {
      wrapper.trigger('keydown', {
        key: 'a'
      });
      await wrapper.vm.$nextTick();
  
      wrapper.trigger('keydown', {
        key: 'Esc',
      });
      await wrapper.vm.$nextTick();
  
      wrapper.trigger('keydown', {
        key: 'Escape',
      });
      await wrapper.vm.$nextTick();
  
      expect(wrapper.emitted().closed.length).toBe(2);
    });

    // describe('after removeNoScroll has been called', ()=>{
    //   beforeEach(()=>{
    //     wrapper.vm.removeNoScroll();
    //   });

    //   it('renders correctly', () => {
    //     expect(wrapper.element).toMatchSnapshot();
    //   });

    //   it('does NOT contain the "cdr-modal__noscroll" class', () => {
    //     expect(document.documentElement.classList.contains('cdr-modal__noscroll')).toBeFalsy();
    //     expect(document.body.classList.contains('cdr-modal__noscroll')).toBeFalsy();
    //   });
    // });
  });


  describe('default closed', ()=>{
    let wrapper;
    let elem;
    beforeEach(()=>{
      elem = document.createElement('div')
      if (document.body) {
        document.body.appendChild(elem)
      }
      wrapper = mount(CdrModal, {
        propsData: {
          opened: false,
          label: "Label is the modal title"
        },
        slots: {
          default: 'Sticky content',
        },
        attachTo: elem,
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('fullscreen snapshot', ()=>{
    let wrapper;
    let elem;
    beforeEach(()=>{
      elem = document.createElement('div')
      if (document.body) {
        document.body.appendChild(elem)
      }
      wrapper = mount(CdrModal, {
        propsData: {
          opened: true,
          label: "Label is the modal title"
        },
        data() {
          return {
            fullscreen: true,
          };
        },
        slots: {
          scrollingContentSlot: 'Main content',
        },
        attachTo: elem,
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
