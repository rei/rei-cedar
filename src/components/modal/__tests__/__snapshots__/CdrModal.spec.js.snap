// Vitest Snapshot v1

exports[`CdrModal.vue > default closed > renders correctly 1`] = `
<div
  class="cdr-modal cdr-modal--closed"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <!-- This div (and the one below) is used to avoid a screen reader "keyboard" trap where the
           content outside the modal is not properly obscured when opened using certain readers.
           Once more browsers catch up to the spec, this hack can probably be removed.
           https://a11ysupport.io/tests/apg__modal-dialog-example -->
    <div />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <!-- @slot Use to override the entire CdrModal content container.
          You must provide an explicit way to close the modal
          to meet UX and accessibility standards  -->
      
      <div
        class="cdr-modal__innerWrap"
        style="display: none;"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <!-- @slot Use to override the default title -->
                
                <h1>
                  Label is the modal title
                </h1>
                <!-- @slot CdrModal content -->
                
              </div>
              <button
                aria-label="Close"
                class="cdr-button cdr-button--primary cdr-button--icon-only-medium cdr-button--icon-only cdr-button--with-background cdr-modal__close-button"
                type="button"
              >
                <!-- @slot Icon to the left of text content -->
                
                
                <!-- @slot Icon for icon-only button -->
                
                <svg
                  aria-hidden="true"
                  class="cdr-icon cdr-icon--medium cdr-icon--inherit-color"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
                  
                  
                  <path
                    d="M13.415 12.006l5.295-5.292a1 1 0 00-1.414-1.415L12 10.591 6.71 5.296A1 1 0 005.295 6.71l5.292 5.295-5.295 5.292a1 1 0 101.414 1.414l5.295-5.292 5.292 5.295a1 1 0 001.414-1.414l-5.292-5.294z"
                    role="presentation"
                  />
                  
                  <!--v-if-->
                </svg>
                
                <!-- @slot Readable text of the button. Leave empty if icon-only -->
                
                
                <!-- @slot Icon to the right of text content -->
                
                
              </button>
            </div>
            <div
              class="cdr-modal__text-content"
              role="document"
              style="max-height: -80px; padding-right: 0px;"
              tabindex="0"
            >
              
              Sticky content
              
            </div>
          </div>
        </section>
      </div>
      
    </div>
    <div />
  </div>
</div>
`;

exports[`CdrModal.vue > default open > renders correctly 1`] = `
<div
  class="cdr-modal"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <!-- This div (and the one below) is used to avoid a screen reader "keyboard" trap where the
           content outside the modal is not properly obscured when opened using certain readers.
           Once more browsers catch up to the spec, this hack can probably be removed.
           https://a11ysupport.io/tests/apg__modal-dialog-example -->
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <!-- @slot Use to override the entire CdrModal content container.
          You must provide an explicit way to close the modal
          to meet UX and accessibility standards  -->
      
      <div
        class="cdr-modal__innerWrap"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <!-- @slot Use to override the default title -->
                
                <h1>
                  Label is the modal title
                </h1>
                <!-- @slot CdrModal content -->
                
              </div>
              <button
                aria-label="Close"
                class="cdr-button cdr-button--primary cdr-button--icon-only-medium cdr-button--icon-only cdr-button--with-background cdr-modal__close-button"
                type="button"
              >
                <!-- @slot Icon to the left of text content -->
                
                
                <!-- @slot Icon for icon-only button -->
                
                <svg
                  aria-hidden="true"
                  class="cdr-icon cdr-icon--medium cdr-icon--inherit-color"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
                  
                  
                  <path
                    d="M13.415 12.006l5.295-5.292a1 1 0 00-1.414-1.415L12 10.591 6.71 5.296A1 1 0 005.295 6.71l5.292 5.295-5.295 5.292a1 1 0 101.414 1.414l5.295-5.292 5.292 5.295a1 1 0 001.414-1.414l-5.292-5.294z"
                    role="presentation"
                  />
                  
                  <!--v-if-->
                </svg>
                
                <!-- @slot Readable text of the button. Leave empty if icon-only -->
                
                
                <!-- @slot Icon to the right of text content -->
                
                
              </button>
            </div>
            <div
              class="cdr-modal__text-content"
              role="document"
              style="max-height: 688px; padding-right: 0px;"
              tabindex="0"
            >
              
              Sticky content
              
            </div>
          </div>
        </section>
      </div>
      
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`CdrModal.vue > fullscreen snapshot > renders correctly 1`] = `
<div
  class="cdr-modal"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <!-- This div (and the one below) is used to avoid a screen reader "keyboard" trap where the
           content outside the modal is not properly obscured when opened using certain readers.
           Once more browsers catch up to the spec, this hack can probably be removed.
           https://a11ysupport.io/tests/apg__modal-dialog-example -->
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <!-- @slot Use to override the entire CdrModal content container.
          You must provide an explicit way to close the modal
          to meet UX and accessibility standards  -->
      
      <div
        class="cdr-modal__innerWrap"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <!-- @slot Use to override the default title -->
                
                <h1>
                  Label is the modal title
                </h1>
                <!-- @slot CdrModal content -->
                
              </div>
              <button
                aria-label="Close"
                class="cdr-button cdr-button--primary cdr-button--icon-only-medium cdr-button--icon-only cdr-button--with-background cdr-modal__close-button"
                type="button"
              >
                <!-- @slot Icon to the left of text content -->
                
                
                <!-- @slot Icon for icon-only button -->
                
                <svg
                  aria-hidden="true"
                  class="cdr-icon cdr-icon--medium cdr-icon--inherit-color"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
                  
                  
                  <path
                    d="M13.415 12.006l5.295-5.292a1 1 0 00-1.414-1.415L12 10.591 6.71 5.296A1 1 0 005.295 6.71l5.292 5.295-5.295 5.292a1 1 0 101.414 1.414l5.295-5.292 5.292 5.295a1 1 0 001.414-1.414l-5.292-5.294z"
                    role="presentation"
                  />
                  
                  <!--v-if-->
                </svg>
                
                <!-- @slot Readable text of the button. Leave empty if icon-only -->
                
                
                <!-- @slot Icon to the right of text content -->
                
                
              </button>
            </div>
            <div
              class="cdr-modal__text-content"
              role="document"
              style="max-height: 688px; padding-right: 0px;"
              tabindex="0"
            >
              
              
            </div>
          </div>
        </section>
      </div>
      
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;
